### PR TITLE
Make text loaders consistent

### LINKED ISSUE #2472: Any reason for inconsistency in params for TextLoader?
1) This Method :
https://github.com/dotnet/machinelearning/blob/834e4715afa070f5e5c3628693682943b3f0bd6d/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs#L55

2) and this one : 
https://github.com/dotnet/machinelearning/blob/834e4715afa070f5e5c3628693682943b3f0bd6d/src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs#L22

Looks like the first one has params like : 
```
char separatorChar = '\t', bool allowQuotedStrings = true, bool supportSparse = true, bool trimWhitespace = false
```

Where as the second one doesn't have those. In order to use those params there is another method overload : 
```
CreateTextLoader(this DataOperationsCatalog catalog, TextLoader.Arguments args, IMultiStreamSource dataSample = null);
```

This takes Arguments class and particularly one param is interesting in this Arguments:
```
char [] Separators
```
 Any reason this is char[] and not char like in other overloads? and why inconsistency in params for the methods 1) and 2).

### PR DESCRIPTION
**Pull request by wschin**
Another attempt to fix #2472.

### CODE REVIEW

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 1]
```diff
@@ -20,15 +20,17 @@ public static class TextLoaderSaverCatalog
...
+        /// <param name="trimWhitespace">Remove trailing whitespace from lines</param>
         /// <param name="dataSample">The optional location of a data sample. The sample can be used to infer column names and number of slots in each column.</param>
         public static TextLoader CreateTextLoader(this DataOperationsCatalog catalog,
             TextLoader.Column[] columns,
             char separatorChar = TextLoader.Defaults.Separator,
             bool hasHeader = TextLoader.Defaults.HasHeader,
             bool allowSparse = TextLoader.Defaults.AllowSparse,
             bool allowQuoting = TextLoader.Defaults.AllowQuoting,
+            bool trimWhitespace = TextLoader.Defaults.TrimWhitespace,
             IMultiStreamSource dataSample = null)
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: The data sample is ***really important***. (See #2663.) If there are feature names, if there is a header, that's where they're coming from. It's really important. Burying it... what... *eight* levels deep seems a bit much. To bury it beyond such things as allowing sparsity and quoting itself seems really tough.

I might say the following order is most sensible:

1. Columns,
2. Separator char,
3. Has header,
4. Data sample,
5. Quoting,
6. Trimming,
7. Sparsity.

I might even think that 5, 6, and 7 could be relegated to advanced options, but no need to insist.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 2]
```diff
@@ -47,71 +49,78 @@ public static class TextLoaderSaverCatalog
...
         /// <param name="allowQuoting">Whether the input may include quoted values,
         /// which can contain separator characters, colons,
         /// and distinguish empty values from missing values. When true, consecutive separators
         /// denote a missing value and an empty value is denoted by \"\".
         /// When false, consecutive separators denote an empty value.</param>
-        /// <param name="allowSparse">Whether the input may include sparse representations for example,
-        /// if one of the row contains "5 2:6 4:3" that's mean there are 5 columns all zero
-        /// except for 3rd and 5th columns which have values 6 and 3</param>
         /// <param name="trimWhitespace">Remove trailing whitespace from lines</param>
         public static TextLoader CreateTextLoader<TInput>(this DataOperationsCatalog catalog,
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: >CreateTextLoader [](start = 33, length = 16)

This is not consistent. The file is missing altogether, which means it cannot be used to read slot names. Whoops.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs [hunk 3]
```diff
@@ -1069,17 +1069,19 @@ private bool HasHeader
         /// <param name="allowSparse">Whether the file can contain numerical vectors in sparse format.</param>
         /// <param name="allowQuoting">Whether the content of a column can be parsed from a string starting and ending with quote.</param>
         /// <param name="dataSample">Allows to expose items that can be used for reading.</param>
+        /// <param name="trimWhitespace">Remove trailing whitespace from lines.</param>
         internal TextLoader(IHostEnvironment env, Column[] columns, char separatorChar = Defaults.Separator,
             bool hasHeader = Defaults.HasHeader, bool allowSparse = Defaults.AllowSparse,
-            bool allowQuoting = Defaults.AllowQuoting, IMultiStreamSource dataSample = null)
-            : this(env, MakeArgs(columns, hasHeader, new[] { separatorChar }, allowSparse, allowQuoting), dataSample)
+            bool allowQuoting = Defaults.AllowQuoting, IMultiStreamSource dataSample = null, bool trimWhitespace = Defaults.TrimWhitespace)
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs]: >IMultiStreamSource dataSample = null [](start = 55, length = 36)

If you wanted you could just get rid of this constructor opportunistically. It serves no purpose as far as I can tell.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 4]
```diff
@@ -47,71 +49,78 @@ public static class TextLoaderSaverCatalog
...
         /// <param name="separatorChar">Column separator character. Default is '\t'</param>
         /// <param name="hasHeader">Does the file contains header?</param>
+        /// <param name="allowSparse">Whether the input may include sparse representations for example,
+        /// if one of the row contains "5 2:6 4:3" that's mean there are 5 columns all zero
+        /// except for 3rd and 5th columns which have values 6 and 3</param>
         /// <param name="allowQuoting">Whether the input may include quoted values,
         /// which can contain separator characters, colons,
         /// and distinguish empty values from missing values. When true, consecutive separators
         /// denote a missing value and an empty value is denoted by \"\".
         /// When false, consecutive separators denote an empty value.</param>
```
Comment (artidoro) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: If you could make the comments consistent as well it would be great.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs (hunk 2 above)
Comment (wschin) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: Ok. I wrongly thought TInput can be used to encode SlotNames.

---
In reply to: [259963380](https://github.com/dotnet/machinelearning/pull/2710#discussion_r259963380) [](ancestors = 259963380)

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs (hunk 1 above)
Comment (wschin) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: Just got them reordered. Thanks for sharing the list!

---
In reply to: [259963185](https://github.com/dotnet/machinelearning/pull/2710#discussion_r259963185) [](ancestors = 259963185)

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs (hunk 4 above)
Comment (wschin) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: Sure.

---
In reply to: [259975940](https://github.com/dotnet/machinelearning/pull/2710#discussion_r259975940) [](ancestors = 259975940)

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs (hunk 3 above)
Comment (wschin) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoader.cs]: Dropped.

---
In reply to: [259963695](https://github.com/dotnet/machinelearning/pull/2710#discussion_r259963695) [](ancestors = 259963695)

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs (hunk 1 above)
Comment (yaeldekel) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: I think that if hasHeader is true, then the dataSample should not be allowed to be null (because if it is, there can be no slot names). Does it make sense to check this? Another alternative is instead of having a bool argument, to have an `IMultiStreamSource header` instead. But on the other hand, if there are two `IMultiStreamSource`s that might be confusing. Not sure what the right signature is here. @ivanidzo4ka, @TomFinley, what do you think?

---
In reply to: [260017117](https://github.com/dotnet/machinelearning/pull/2710#discussion_r260017117) [](ancestors = 260017117,259963185)

Comment (TomFinley) [reply src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: > I think that if hasHeader is true, then the dataSample should not be allowed to be null (because if it is, there can be no slot names). Does it make sense to check this? Another alternative is instead of having a bool argument, to have an `IMultiStreamSource header` instead. But on the other hand, if there are two `IMultiStreamSource`s that might be confusing. Not sure what the right signature is here. @Ivanidzo4ka, @TomFinley, what do you think?
> 
> In reply to: [260017117](https://github.com/dotnet/machinelearning/pull/2710#discussion_r260017117) [](ancestors = 260017117,259963185)

So, let's imagine that I declare that there is a header, but I provide no sample. I think this is fine. I am merely describing *how* to read data, and I am saying that when you read the data with this thing, that you will have a header to contend with. If headers were always only used for defining slot names, I might agree, but sometimes files have headers just because they have headers, and there's no larger reason for it.

The presence of absence of a sample does not mean that we cannot fully define the schema, even if header is true. (This is in contrast to some other options, like, "this is fixed size, but I would like you to auto-determine what that size is, which obviously requires a sample.)

So no check would be my reaction.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 5]
```diff
@@ -47,79 +61,98 @@ public static class TextLoaderSaverCatalog
         /// <param name="catalog">The <see cref="DataOperationsCatalog"/> catalog.</param>
         /// <param name="separatorChar">Column separator character. Default is '\t'</param>
         /// <param name="hasHeader">Does the file contains header?</param>
```
Comment (singlis) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: >hasHeader [](start = 25, length = 9)

Nit pick - I would rephrase this as "Set to true if the file contains a header row".

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 6]
```diff
@@ -128,14 +161,19 @@ public static class TextLoaderSaverCatalog
         /// <param name="catalog">The <see cref="DataOperationsCatalog"/> catalog.</param>
         /// <param name="path">Specifies a file from which to read.</param>
         /// <param name="options">Defines the settings of the load operation.</param>
-        public static IDataView ReadFromTextFile(this DataOperationsCatalog catalog, string path, TextLoader.Options options = null)
+        /// <param name="dataSample">The optional location of a data sample. The sample can be used to infer column names and number of slots in each column.</param>
+        public static IDataView ReadFromTextFile(this DataOperationsCatalog catalog, string path,
+            TextLoader.Options options = null, IMultiStreamSource dataSample = null)
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: >, IMultiStreamSource dataSample = null [](start = 45, length = 38)

Hi @wschin, somehow I missed this -- definitely we don't need to have a data sample if we're talking about the path to a data file already.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 7]
```diff
@@ -47,79 +61,98 @@ public static class TextLoaderSaverCatalog
...
         /// if one of the row contains "5 2:6 4:3" that's mean there are 5 columns all zero
         /// except for 3rd and 5th columns which have values 6 and 3</param>
-        /// <param name="trimWhitespace">Remove trailing whitespace from lines</param>
-        /// <param name="path">The path to the file.</param>
         /// <returns>The data view.</returns>
         public static IDataView ReadFromTextFile<TInput>(this DataOperationsCatalog catalog,
             string path,
             char separatorChar = TextLoader.Defaults.Separator,
             bool hasHeader = TextLoader.Defaults.HasHeader,
+            IMultiStreamSource dataSample = null,
```
Comment (TomFinley) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: >IMultiStreamSource dataSample [](start = 12, length = 29)

Similar here.

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 8]
```diff
@@ -47,79 +61,98 @@ public static class TextLoaderSaverCatalog
...
             bool allowQuoting = TextLoader.Defaults.AllowQuoting,
-            bool allowSparse = TextLoader.Defaults.AllowSparse,
-            bool trimWhitespace = TextLoader.Defaults.TrimWhitespace)
-            => TextLoader.CreateTextReader<TInput>(CatalogUtils.GetEnvironment(catalog), hasHeader, separatorChar, allowQuoting, allowSparse, trimWhitespace);
+            bool trimWhitespace = TextLoader.Defaults.TrimWhitespace,
+            bool allowSparse = TextLoader.Defaults.AllowSparse)
+            => TextLoader.CreateTextReader<TInput>(CatalogUtils.GetEnvironment(catalog), hasHeader, separatorChar, allowQuoting,
+                allowSparse, trimWhitespace, dataSample: dataSample);
```
Comment (singlis) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: >dataSample [](start = 45, length = 10)

why pass the arguments this way? Why not add dataSample after hasHeader?

File: src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs [hunk 9]
```diff
@@ -47,79 +61,98 @@ public static class TextLoaderSaverCatalog
...
+        /// <param name="path">The path to the file.</param>
         /// <param name="separatorChar">Column separator character. Default is '\t'</param>
+        /// <param name="hasHeader">Does the file contains header?</param>
+        /// <param name="dataSample">The optional location of a data sample. The sample can be used to infer column names and number of slots in each column.</param>
         /// <param name="allowQuoting">Whether the input may include quoted values,
         /// which can contain separator characters, colons,
         /// and distinguish empty values from missing values. When true, consecutive separators
         /// denote a missing value and an empty value is denoted by \"\".
         /// When false, consecutive separators denote an empty value.</param>
+        /// <param name="trimWhitespace">Remove trailing whitespace from lines</param>
```
Comment (singlis) [inline src/Microsoft.ML.Data/DataLoadSave/Text/TextLoaderSaverCatalog.cs]: minor nit - Add a period.
